internal/controller/accessrequest_controller.go

```diff
@@ -33,30 +38,241 @@ type AccessRequestReconciler struct {
 	Scheme *runtime.Scheme
 }
 
+const (
+	// AccessRequestFinalizerName defines the name of the AccessRequest finalizer
+	// managed by this controller
+	AccessRequestFinalizerName = "accessrequest.ephemeral-access.argoproj-labs.io/finalizer"
+)
+
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the AccessRequest object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
+// 1. verify if accessrequest is expired and status is "granted"
+// 1.1 if so, remove the user from the elevated role
+// 1.2 update the accessrequest status to "expired"
+// 1.3 return
+// 2. verify if user has the necessary access to be promoted
+// 2.1 if they don't, update the accessrequest status to "denied"
+// 2.2 return
+// 3. verify if CR is approved
+// 4. retrieve the Application
+// 5. retrieve the AppProject
+// 6. assign user in the desired role in the AppProject
+// 7. update the accessrequest status to "granted"
+// 8. set the RequeueAfter in Result
 func (r *AccessRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx).
+		WithValues("controller", "AccessRequest", "reconcile", req.NamespacedName)
+
+	logger.Info("retrieving AccessRequest k8s state")
+	var ar api.AccessRequest
+	if err := r.Get(ctx, req.NamespacedName, &ar); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "error retrieving AccessRequest from k8s")
+		return ctrl.Result{}, err
+	}
+
+	// check if the object is being deleted and properly handle it
+	deleted, err := r.handleFinalizer(ctx, &ar)
+	if err != nil {
+		logger.Error(err, "handleFinalizer error")
+		return ctrl.Result{}, fmt.Errorf("error handling finalizer: %w", err)
+	}
+	// stop the reconciliation as the object was deleted
+	if deleted {
+		logger.Info("Object deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// initialize the status if not done yet
+	if ar.Status.Status == "" {
```

Comment:
Can the last `Status` field be renamed? `.Status.Status` is a bit confusing to read. Maybe `RequestState`?

Reply:
done